accepted/2020/binary-data/binary-data.md

```diff
@@ -0,0 +1,144 @@
+        await UploadImageAsync(newUrl, image);
+    }
+}
+```
+
+## Requirements
+
+### Goals
+
+* Stable API
+* Make it easy for people to construct bytes from common data types (byte
+  arrays, spans, memories, string, JSON serialized objects)
+* Make it easy for people to convert bytes to common data types of the same
+  representations
+* Provide opinionated defaults in order to reduce complexity
+    - String encoding is assumed to be UTF8
+    - Object serialization is done with `System.Text.Json`, with specific options
+
+### Non-Goals
+
+* Allocation free zero-copying semantics. The primary goal is convenience and
+  usability, not perf. Performance critical scenarios should be enabled by
+  providing overloads with the existing low-level concepts, such as span.
+
+## Design
+
+We're concerned with hard coding the type to `System.Text.Json`, but we
+acknowledge that defaults are very valuable. If we need to support other
+serializers we can provide additional arguments or APIs.
+
+This type can't live in corlib because it would depend on `System.Text.Json`
+(and whatever future serializers it needs to support). Hence, most areas in the
+BCL wouldn't be able to take `BinaryData`, but technologies above the BCL could
+(Microsoft.Extensions.*, ASP.NET Core, Azure SDK etc). While that might be
+unfortunate, we believe it's acceptable because most of the usability gains
+would be in higher levels anyway.
+
+The type isn't meant to unify different representations. Rather, it's a helper
+to make it easier for users to convert data. Thus, it's not meant as
+high-performance exchange type, like span (it is implicitly convertible to
+`ReadOnlyMemory<byte>` and `ReadOnlySpan<byte>`).
+
+The type is for "user data", that is, the contract for the API is "give me a
+blob of binary data". It's not for protocol-like APIs where the underlying
+encoding is part of the API contract, e.g. a REST API.
+
+We don't see the need to, for example, provide new virtual methods on Stream or
+`HttpClient`.
+
+### Shipping vehicle
+
+**Assembly**: System.BinaryData.dll
+
+**NuGet Package**: System.BinaryData (stable)
+
+### API
+
+```C#
+namespace System
+{
+    public class BinaryData
+    {
+        public BinaryData(byte[] data);
+        public BinaryData(object jsonSerializable, JsonSerializerOptions options = default, Type? type = null);
+        public BinaryData(ReadOnlyMemory<byte> data);
+        public BinaryData(string data);
+        public static BinaryData FromBytes(ReadOnlyMemory<byte> data);
+        public static BinaryData FromBytes(byte[] data);
+        public static BinaryData FromObjectAsJson<T>(T jsonSerializable, JsonSerializerOptions options = default, CancellationToken cancellationToken = default);
+        public static BinaryData FromStream(Stream stream);
+        public static Task<BinaryData> FromStreamAsync(Stream stream, CancellationToken cancellationToken = default);
+        public static BinaryData FromString(string data);
```

Comment:
> String encoding is assumed to be UTF8

What is string in context here -- eg., this string is regular UTF-16. Are you referring to eg., `byte[] data` ? Should UTF-8 appear in any method names?

Reply:
"encoding assumed to be UTF-8" means that any string data provided here will be converted to UTF-8, and the resulting `BinaryData` instance will contain the UTF-8 representation of that string data. I don't think it's necessary to state UTF-8 in any API name.